### ISSUE TITLE
How to run in parallel

### ISSUE THREAD

**Issue by bdeonovic** [raphael-group/multibreak-sv#6]
In the documentation it is suggested that MBSV could be parallelized by using the GASV clusters independently. What needs to be done to parallelize this process? Currently I am running MBSV, but it is taking a very long time (my input m5 file was 80Gb, so might be intractable). 

Also, I would like to combine long and short reads to run MBSV, will there be documentation to do this?

**Comment by annaritz** [comment]
Hi - thanks for your interest.  

To run MBSV in parallel, you can make use of the pre-processor that splits the GASV clusters file into independent sub-problems.  From the manual (under doc/):

(6) SPLITTING ESP FILE
- GASV requires an Lmin/Lmax, which changes depending on the fragment size.
- "Bin" the ESPs into different Lmin/Lmax regions.

Each one of these files can be run independently with MBSV - you can pass it the same assignment file for all runs, and the program will only keep the read alignments that appear in that particular sub-problem.  In the "Example Runs" you simply need to change the --clusterfile argument each time (and of course the --prefix argument to prevent overwriting results).

By short reads, do you mean something like Illumina paired-end reads?  There is some documentation at the end of the manual, and I hope to have a pre-processor/pipeline that makes this easier in a few weeks.  If you mean PacBio long reads with different library lengths (e.g., different PacBio runs), you can simply have different assignment files and have two lines in the experiment file denoting the coverage, error rate, etc. for each run.

If you have more specific questions, feel free to send me an email.  Thanks, Anna

**Comment by bdeonovic** [comment]
Thank you for the comprehensive answer. A few clarifying questions:

When I originally ran MBSV I ran it with the following command:

``` bash
java -Xmx512000m -jar MultiBreakSV.jar --clusterfile out-RunGASV/gasv.in.clusters --assignmentfile out-MBSVinputs/assignments.txt --experimentfile out-MBSVinputs/experiments.txt --numiterations 10000 --perr 0.001
```

also I see the following folder 

``` bash
[bdeonovic@neon-login-0-2 gold_standard]$ ls -lrth out-RunGASV/binned-esps/
total 47M
-rw-r--r-- 1 bdeonovic its-rs-neon    0 Jun 11 18:01 translocations
-rw-r--r-- 1 bdeonovic its-rs-neon 129M Jun 11 18:01 intrachrom-esps-bin-0-199
-rw-r--r-- 1 bdeonovic its-rs-neon  17M Jun 11 18:01 intrachrom-esps-bin-200-399
-rw-r--r-- 1 bdeonovic its-rs-neon 3.3M Jun 11 18:01 intrachrom-esps-bin-400-599
-rw-r--r-- 1 bdeonovic its-rs-neon 1.4M Jun 11 18:01 intrachrom-esps-bin-600-799
-rw-r--r-- 1 bdeonovic its-rs-neon 344K Jun 11 18:01 intrachrom-esps-bin-800-999
-rw-r--r-- 1 bdeonovic its-rs-neon 207K Jun 11 18:01 intrachrom-esps-bin-1000-1199
-rw-r--r-- 1 bdeonovic its-rs-neon 163K Jun 11 18:01 intrachrom-esps-bin-1200-1399
-rw-r--r-- 1 bdeonovic its-rs-neon  66K Jun 11 18:01 intrachrom-esps-bin-1400-1599
-rw-r--r-- 1 bdeonovic its-rs-neon  26K Jun 11 18:01 intrachrom-esps-bin-1600-1799
-rw-r--r-- 1 bdeonovic its-rs-neon  27K Jun 11 18:01 intrachrom-esps-bin-1800-1999
-rw-r--r-- 1 bdeonovic its-rs-neon  14K Jun 11 18:01 intrachrom-esps-bin-2000-2199
-rw-r--r-- 1 bdeonovic its-rs-neon  17K Jun 11 18:01 intrachrom-esps-bin-2200-2399
-rw-r--r-- 1 bdeonovic its-rs-neon 7.7K Jun 11 18:01 intrachrom-esps-bin-2400-2599
-rw-r--r-- 1 bdeonovic its-rs-neon 3.3K Jun 11 18:01 intrachrom-esps-bin-2600-2799
-rw-r--r-- 1 bdeonovic its-rs-neon 1.5K Jun 11 18:01 intrachrom-esps-bin-2800-2999
-rw-r--r-- 1 bdeonovic its-rs-neon  309 Jun 11 18:01 intrachrom-esps-bin-3000-3199
-rw-r--r-- 1 bdeonovic its-rs-neon  242 Jun 11 18:01 intrachrom-esps-bin-3200-3399
-rw-r--r-- 1 bdeonovic its-rs-neon  126 Jun 11 18:01 intrachrom-esps-bin-3400-3599
```

Are these the binned ESPs that you refer to? Are these the files that you run independently with MBSV? Their contents don't seem to match any of the MBSV input files:

``` bash
[bdeonovic@neon-login-0-2 gold_standard]$  head out-RunGASV/binned-esps/intrachrom-esps-bin-0-199 
longread_4957281_0.1-1.1        1       14368   14368   +       1       17056   17056   -
longread_7508524        1       14390   14390   +       1       17685   17685   -
longread_1561209_0.2-1.2        1       14399   14399   +       1       188115  188115  +
longread_6257015_0.6-1.6        1       14411   14411   +       1       186462  186462  -
longread_6257015_0.7-1.7        1       14411   14411   +       1       186462  186462  -
longread_4957284_0.1-1.1        1       14414   14414   +       1       17049   17049   -
longread_3223104_0.5-1.5        1       14552   14552   -       1       187466  187466  -
longread_867689_0.1-1.1 1       14796   14796   -       1       22823790        22823790        +
longread_7398951        1       15795   15795   +       1       17615   17615   -
longread_7398952        1       15796   15796   +       1       20108   20108   -
```

If these are not the files, how do you "split the ESP files", the documentation doesn't mention this (is it already done automatically?)

**Comment by annaritz** [comment]
Hi - yep, you are right.  Sorry, that was a different part of the pipeline where we need to divide the ESPs to determine the proper arguments for clustering.

It appears that I didn't commit the code that splits the clusters file into smaller pieces.  I will do this in the next couple days, email you when it's done, and close this issue.  Thanks for the catch.

**Comment by bdeonovic** [comment]
Just following up on the progress of this commit. Also to clarify: you will be submitting an update with a script that will split cluster files into smaller pieces so that the smaller pieces can be run independently in parallel, is that right?

**Comment by annaritz** [comment]
I just committed a perl script in src/; if you re-install it will be copied to bin/.  This script takes a clusters file and and output directory and splits the clusters into independent subproblems.  For the script to work, your reads need to have a "longread_#" prefix (it looks like yours do).  

To speed up each run for MultiBreak-SV, use the --readclustersfirst option.  See Section IV. in the manual for more details, and let me know if you run into any problems. Just open a new ticket.  Thanks -

**Comment by bdeonovic** [comment]
Thank you for providing this script! The script mentions 3 inputs that are required:

``` perl
### Three arguments are required:
### (1) clusters file (*.clusters)
### (2) batch input file for GASV that lists all ESPs
### (3) output directory.
```

which file is number 2 refer to? After running `M5toMBSV.py` I have the following directory

``` batch
[bdeonovic@neon-login-0-2 gold_standard]$ ls -lrth out-RunGASV/
total 139M
drwxr-xr-x 2 bdeonovic its-rs-neon   21 Jun 11 18:01 binned-esps
-rw-r--r-- 1 bdeonovic its-rs-neon 1.3K Jun 11 18:01 gasv.in
-rw-r--r-- 1 bdeonovic its-rs-neon 422M Jun 11 22:50 gasv.in.clusters
```

I believe the file `gasv.in.clusters` is the file used for (1). The directory `binned-esps` contains files that look like 

``` batch
[bdeonovic@neon-login-0-2 gold_standard]$ ls -lrth out-RunGASV/binned-esps/
total 47M
-rw-r--r-- 1 bdeonovic its-rs-neon    0 Jun 11 18:01 translocations
-rw-r--r-- 1 bdeonovic its-rs-neon 129M Jun 11 18:01 intrachrom-esps-bin-0-199
-rw-r--r-- 1 bdeonovic its-rs-neon  17M Jun 11 18:01 intrachrom-esps-bin-200-399
-rw-r--r-- 1 bdeonovic its-rs-neon 3.3M Jun 11 18:01 intrachrom-esps-bin-400-599
-rw-r--r-- 1 bdeonovic its-rs-neon 1.4M Jun 11 18:01 intrachrom-esps-bin-600-799
-rw-r--r-- 1 bdeonovic its-rs-neon 344K Jun 11 18:01 intrachrom-esps-bin-800-999
-rw-r--r-- 1 bdeonovic its-rs-neon 207K Jun 11 18:01 intrachrom-esps-bin-1000-1199
-rw-r--r-- 1 bdeonovic its-rs-neon 163K Jun 11 18:01 intrachrom-esps-bin-1200-1399
-rw-r--r-- 1 bdeonovic its-rs-neon  66K Jun 11 18:01 intrachrom-esps-bin-1400-1599
-rw-r--r-- 1 bdeonovic its-rs-neon  26K Jun 11 18:01 intrachrom-esps-bin-1600-1799
-rw-r--r-- 1 bdeonovic its-rs-neon  27K Jun 11 18:01 intrachrom-esps-bin-1800-1999
-rw-r--r-- 1 bdeonovic its-rs-neon  14K Jun 11 18:01 intrachrom-esps-bin-2000-2199
-rw-r--r-- 1 bdeonovic its-rs-neon  17K Jun 11 18:01 intrachrom-esps-bin-2200-2399
-rw-r--r-- 1 bdeonovic its-rs-neon 7.7K Jun 11 18:01 intrachrom-esps-bin-2400-2599
-rw-r--r-- 1 bdeonovic its-rs-neon 3.3K Jun 11 18:01 intrachrom-esps-bin-2600-2799
-rw-r--r-- 1 bdeonovic its-rs-neon 1.5K Jun 11 18:01 intrachrom-esps-bin-2800-2999
-rw-r--r-- 1 bdeonovic its-rs-neon  309 Jun 11 18:01 intrachrom-esps-bin-3000-3199
-rw-r--r-- 1 bdeonovic its-rs-neon  242 Jun 11 18:01 intrachrom-esps-bin-3200-3399
-rw-r--r-- 1 bdeonovic its-rs-neon  126 Jun 11 18:01 intrachrom-esps-bin-3400-3599
```

**Comment by annaritz** [comment]
Hi - Sorry, that comment was old! I fixed it - it just takes a clusters file and an output directory now, assuming that your multi-breakpoint-read naming scheme is from M5toMBVS.py (e.g., longread_12345).  The running instructions and examples are in the Manual (under doc/)
